REGRESSION (278067@main): UI process crashes under immediateActionRecognizerWillPrepare if the Web Content has crashed before you click.

megangardner opened the pull request:
#### d48763c9c2dea7e864c05a2731e910e8387fcd67
<pre>
Fix for UIProcess crash after 273330.
<a href="https://bugs.webkit.org/show_bug.cgi?id=273663">https://bugs.webkit.org/show_bug.cgi?id=273663</a>
<a href="https://rdar.apple.com/127466607">rdar://127466607</a>

Reviewed by Abrar Rahman Protyasha, Tim Horton and Wenson Hsieh.

If the web process crashes, then this UI process will crash with no
mainFrame and/or frameID.

* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController immediateActionRecognizerWillPrepare:]):

Canonical link: <a href="https://commits.webkit.org/278323@main">https://commits.webkit.org/278323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fba4e3a7cf513ec4f522b56bdb88751957cf5c38

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50130 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29421 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53388 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/820 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35616 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40913 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52229 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22005 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8507 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/452 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54969 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25226 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/418 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48308 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26487 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47320 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27350 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7253 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26218 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
Nit: squash the two if conditions with an|| operator

hortont424 commented on Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
```diff
@@ -175,6 +175,12 @@ - (void)immediateActionRecognizerWillPrepare:(NSImmediateActionGestureRecognizer
+    if (!_page->mainFrame())
+        return;
+
+    if (!_page->mainFrame()->frameID())
```
This one's just an ID so I don't think you have to check it? Also, I don't think you'll have a non null frame without an ID anyway.

whsieh replied:
I think a 0 `frameID` would fail at decoding time — but yeah, it does seem like this check is unnecessary though, since a non-null frame implies non-zero ID.

hortont424 commented on Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
```diff
@@ -175,6 +175,12 @@ - (void)immediateActionRecognizerWillPrepare:(NSImmediateActionGestureRecognizer
     _state = WebKit::ImmediateActionState::Pending;
     immediateActionRecognizer.animationController = nil;
 
+    if (!_page->mainFrame())
```
Should we bail earlier? Also, if we do, I think we need to keep a smart pointer of whatever kind on the stack, per the new rules.

(As it stands it's ok since you're only using it for a trivial getter but if you move the check up higher the object could disappear between then and the use below.)

megangardner replied:
I thought about bailing earlier, but I wasn't sure what the standard was.

megangardner replied:
If we don't _have_ to check earlier, it makes sense to me to just do it here, but I will change it if required.

megangardner committed:
Fix for UIProcess crash after 273330.
https://bugs.webkit.org/show_bug.cgi?id=273663
rdar://127466607

Reviewed by Abrar Rahman Protyasha, Tim Horton and Wenson Hsieh.

If the web process crashes, then this UI process will crash with no
mainFrame and/or frameID.

* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController immediateActionRecognizerWillPrepare:]):

Canonical link: https://commits.webkit.org/278323@main
